### PR TITLE
Add alicloud provider, including resource and examples of alicloud provider

### PR DESCRIPTION
**Pull request by xiaozhu36**
This is terraform alicloud provider provided by Official AliCloud. The provider is aim at building, changing, and managing alicloud infrastructure safely and efficiently based on terraform powerful management capabilities for infrastructure. At present, it has supported several alicloud resource, such as ECS, VPC, SLB, Security Group and EIP. With the continuous development and improvement to the provider, it will have more perfect function and support more alicloud resource. We hope the PR to integrate alicloud provider into terraform and help consumer to manage alicloud infrastructure. If you have any question, we are glad to get feedback as soon as possible, thanks a lot.

### CODE REVIEW

File: builtin/providers/alicloud/resource_alicloud_nat_gateway.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+				Optional: true,
+			},
+
+			"bandwidth_package_ids": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"bandwidth_packages": &schema.Schema{
+				Type: schema.TypeList,
```
Comment (stack72) [inline builtin/providers/alicloud/resource_alicloud_nat_gateway.go]: As I see only 4 `bandwidth_packages` are allowed, we can add:

`MaxItems: 4` to the schema

We can then remove the if len() > 4 check below

File: builtin/providers/alicloud/resource_alicloud_nat_gateway.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+		bandwidthPackageTypes = append(bandwidthPackageTypes, bandwidthPackage)
+	}
+
+	args.BandwidthPackage = bandwidthPackageTypes
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	}
+	// } else {
```
Comment (stack72) [inline builtin/providers/alicloud/resource_alicloud_nat_gateway.go]: Can this be removed or is it a todo?

File: builtin/providers/alicloud/resource_alicloud_nat_gateway.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+	}
+
+	d.SetId(resp.NatGatewayId)
+	d.Partial(true)
+	d.SetPartial("name")
+	d.SetPartial("description")
+	d.SetPartial("spec")
+	d.SetPartial("vpc_id")
+
+	// for i, packageId := range resp.BandwidthPackageIds.BandwidthPackageId {
```
Comment (stack72) [inline builtin/providers/alicloud/resource_alicloud_nat_gateway.go]: redundant comment?

File: builtin/providers/alicloud/resource_alicloud_nat_gateway.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+		args.Description = v.(string)
+	}
+
+	resp, err := CreateNatGateway(conn, args)
+	if err != nil {
+		return fmt.Errorf("CreateNatGateway got error: %#v", err)
+	}
+
+	d.SetId(resp.NatGatewayId)
+	d.Partial(true)
```
Comment (stack72) [inline builtin/providers/alicloud/resource_alicloud_nat_gateway.go]: We don't need d.Partial in Create

File: builtin/providers/alicloud/resource_alicloud_nat_gateway.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+	if d.HasChange("spec") {
+		var spec NatGatewaySpec
+		if v, ok := d.GetOk("spec"); ok {
+			spec = NatGatewaySpec(v.(string))
+		} else {
+			// set default to small spec
+			spec = NatGatewaySmallSpec
+		}
+
+		args := &ModifyNatGatewaySpecArgs{
```
Comment (stack72) [inline builtin/providers/alicloud/resource_alicloud_nat_gateway.go]: Follow the pattern of a single conn.ModifyNatGatewaySpec and pass the args in

**Comment by xiaozhu36** [comment]
Hi @stack72 

Thanks very much for your help and review advise, and our team have improve existing defects in my provider as per your advise. 

Now, we want to pull a new request and the PR has the following changes:

1. Improved the existing defects and fixed some bugs, including removing d.SetPartial in the Create funcs, improving Updatefunc, adding some attributes for the some parameters and so on.

2. Using `govendor` added the SDK correctly and updated the dependencies. 

3. Ran commands 'make fmt' and 'make test vet', and passed it successfully. 

Please review them and looking forward to a good news. Thanks very much.

Thanks

Guimin He

**Comment by stack72** [comment]
Closed via #11235

**Pull request closed without merging**